lib/engine.js

```diff
@@ -0,0 +1,166 @@
+const { createStore, combineReducers } = require('redux');
+const _ = require('lodash');
+
+let store;
+
+// actions
+
+const PICK_OFFENSE_CARD = 'PICK_OFFENSE_CARD';
+const PICK_DEFENSE_CARD = 'PICK_DEFENSE_CARD';
+
+function pickOffenseCard(cardId) {
+  return {
+    type: PICK_OFFENSE_CARD,
+    cardId,
+    player: 'player',
+  };
+}
+
+function pickDefenseCard(cardId) {
+  return {
+    type: PICK_DEFENSE_CARD,
+    cardId,
+    player: 'player',
+  };
+}
+
+// reducers
+
+// game engine reducer helpers
+
+function validateCardChoice(choices, cardId) {
+  return choices && choices[cardId] ? choices[cardId] : null;
+}
+
+function removeFromDeck({ state, action, card }) {
+  const deckKey = action.player === 'ai' ? 'aiDeck' : 'playerDeck';
+  const cardInDeck = state[deckKey][action.cardId];
+  // if this is the last copy of the card, delete it from the deck
+  const updatedState = cardInDeck.count <= 1 ? {
+    ...state,
+    [deckKey]: _.omit(state[deckKey], action.cardId),
+  } : {
+    ...state,
+    [deckKey]: {
+      ...state[deckKey],
+      [action.cardId]: {
+        ...card,
```

Comment:
Ok, so I think here's where I'm getting confused.  Why have `{ state, action, card }` when you have `state[deckKey][action.cardId]`?  Why do you need the card object to be passed in separately?  Oh, is it just because you need to remove it from the deck but then still have a copy of it for when you play it?  Hmm.